services/continuousquery/service.go

```diff
@@ -0,0 +1,285 @@
+		now = now.In(cq.q.Location)
+	}
+
+	// set cq.LastRun and cq.HasRun
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	cq.LastRun, cq.HasRun = s.lastRuns[cqi.Name]
+
+	// If no rp is specified, use the default one.
+	if cq.q.Target.Measurement.RetentionPolicy == "" {
+		cq.q.Target.Measurement.RetentionPolicy = dbi.DefaultRetentionPolicy
+	}
+
+	// Get the group by interval and report time for cq service.
+	interval, err := cq.q.GroupByInterval()
+	if err != nil || interval == 0 {
+		return false, err
+	}
+	// if interval < 5 minutes, set s.ReportInterval to 5 minutes.
+	if interval < DefaultReportTime {
+		s.ReportInterval = DefaultReportTime
+	} else {
+		s.ReportInterval = interval
+	}
+
+	// Get the group by offset.
+	offset, err := cq.q.GroupByOffset()
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the CQ should be run.
+	ok, nextRun, err := cq.shouldRunContinuousQuery(now, interval)
+	if err != nil {
+		return false, err
+	} else if !ok {
+		return false, nil
+	}
+
+	// update cq.LastRun and s.lastRuns
+	cq.LastRun = nextRun
+	s.lastRuns[cqi.Name] = cq.LastRun
+
+	// Calculate and set the time range for the query.
+	// startTime should be earlier than current time.
+	startTime := nextRun.Add(-interval - offset - 1).Truncate(interval).Add(offset)
+	endTime := startTime.Add(interval - offset).Truncate(interval).Add(offset)
+
+	if err := cq.q.SetTimeRange(startTime, endTime); err != nil {
+		return false, fmt.Errorf("unable to set time range: %s", err)
+	}
+
+	// execute the query and write the results
+	res := s.runContinuousQueryAndWriteResult(cq)
+	if res.Err != nil {
+		return false, res.Err
+	}
+
+	return true, nil
+}
+
+// NewContinuousQuery returns a ContinuousQuery object with a parsed SQL statement.
+func NewContinuousQuery(database string, cqi *meta.ContinuousQueryInfo) (*ContinuousQuery, error) {
+	// Parse the query.
+	p := influxql.NewParser(strings.NewReader(cqi.Query))
+	defer p.Release()
+
+	YyParser := influxql.NewYyParser(p.GetScanner(), p.GetPara())
+	YyParser.ParseTokens()
+
+	qr, err := YyParser.GetQuery()
+	if err != nil {
+		return nil, err
+	}
+	if len(qr.Statements) == 0 {
+		return nil, nil
+	}
+	stmt := qr.Statements[0]
+
+	// Check if the statement is a valid continuous query.
+	q, ok := stmt.(*influxql.CreateContinuousQueryStatement)
+	if !ok || q.Source.Target == nil || q.Source.Target.Measurement == nil {
+		// q.Source.Target (destination) for the result of a SELECT INTO query.
+		return nil, errors.New("invalid continuous query")
+	}
+
+	// Create a new ContinuousQuery object.
+	cq := &ContinuousQuery{
+		Database: database,
+		Info:     cqi,
+		q:        q.Source,
+	}
+
+	return cq, nil
+}
+
+// shouldRunContinuousQuery returns true if the CQ should run.
+func (cq *ContinuousQuery) shouldRunContinuousQuery(now time.Time, interval time.Duration) (bool, time.Time, error) {
+	// check whether the query is an aggregate query.
+	if cq.q.IsRawQuery { // cq.q.IsRawQuery is true when is not an aggregate query.
```

Comment:
Whether cq is an aggregate query should be judged when creating. You just care about the time.